cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,380 @@
+/*
+Copyright 2020 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/digitalocean/godo"
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	// Interval of synchronizing service status from apiserver.
+	serviceSyncPeriod = 35 * time.Second
+	// Frequency at which the firewall controller runs.
+	firewallReconcileFrequency = 5 * time.Minute
+)
+
+var (
+	// Outbound rules that we commonly use for our internal firewalls, it is what we should
+	// use for public access firewalls too.
+	allowAllOutboundRules = []godo.OutboundRule{
+		{
+			Protocol:  "tcp",
+			PortRange: "all",
+			Destinations: &godo.Destinations{
+				Addresses: []string{"0.0.0.0/0", "::/0"},
```

Comment:
Out of curiosity: do we need to define those, or does the API possibly do the right thing if we leave the destination empty / nil?

Reply:
Unfortunately, no.

```
"You must specify a list of destinations for an outbound rule."
```